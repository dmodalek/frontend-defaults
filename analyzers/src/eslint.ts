import { join } from 'path';
import { Analyzer, fileExists } from '@namics/frontend-defaults-core';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: false | string;
};

export class ESLintAnalyzer extends Analyzer<ESLintAnalyzerResult> {
	async analyze(): Promise<ESLintAnalyzerResult> {
		const doesESLintExist = await fileExists(join(this.context, '.eslintrc.js'));

		if (doesESLintExist) {
			const doesESLintIgnoreExist = await fileExists(join(this.context, '.eslintignore'));
			const eslintInstallation = this.packageAnalyzer.anyDependencyExists('eslint');

			return {
				eslint: doesESLintExist,
				eslintIgnore: doesESLintIgnoreExist,
				eslintInstallation: eslintInstallation ? eslintInstallation.version : false,
			};
		}

		return {
			eslint: doesESLintExist,
		};
	}
}
