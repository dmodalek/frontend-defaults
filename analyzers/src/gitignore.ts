import { join } from 'path';
import { Analyzer, fileExists } from '@namics/frontend-defaults-core';

export type GitignoreAnalyzerResult = {
	gitignore: boolean;
};

export class GitignoreAnalzyer extends Analyzer<GitignoreAnalyzerResult> {
	async analyze(): Promise<GitignoreAnalyzerResult> {
		const doesGitignoreExist = await fileExists(join(this.context, '.gitignore'));

		return {
			gitignore: doesGitignoreExist,
		};
	}
}
