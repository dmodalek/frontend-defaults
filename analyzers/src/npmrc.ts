import { join } from 'path';
import { Analyzer, fileExists, getFileContents } from '@namics/frontend-defaults-core';

export type NPMRCAnalyzerResult = {
	npmrc: boolean;
	npmrcSaveExactEnabled?: boolean;
};

export class NPMRCAnalyzer extends Analyzer<NPMRCAnalyzerResult> {
	async analyze(): Promise<NPMRCAnalyzerResult> {
		const npmrcPath = join(this.context, '.npmrc');
		const npmrcExists = await fileExists(npmrcPath);

		if (npmrcExists) {
			const contents = await getFileContents(npmrcPath);
			const saveExactEnabled = /^(save-exact)\s?=\s?true$/gim.test(contents);

			return {
				npmrc: npmrcExists,
				npmrcSaveExactEnabled: saveExactEnabled,
			};
		}

		return {
			npmrc: npmrcExists,
		};
	}
}
