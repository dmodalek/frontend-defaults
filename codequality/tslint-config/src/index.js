const fs = require('fs');
const rimraf = require('rimraf');
const rules = require('./rules');
const jsRules = require('./jsRules');
const reactRules = require('./reactRules');

rimraf.sync('./build');
fs.mkdirSync('./build');

fs.writeFileSync(
	'./build/tslint.json',
	JSON.stringify(
		{
			defaultSeverity: 'error',
			extends: ['tslint-react'],
			jsRules,
			rules: Object.assign({}, rules, reactRules),
			rulesDirectory: [],
		},
		null,
		2
	)
);
console.log('wrote file build/tslint.json');
