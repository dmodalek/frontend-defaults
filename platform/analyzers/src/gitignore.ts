import { fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export const gitIgnoreAnalyzer = async (cwd: string): Promise<GitIgnoreAnalyzerResult> => {
	const doesGitignoreExist = await fileExists(join(cwd, '.gitignore'));

	return {
		gitignore: doesGitignoreExist,
	};
}