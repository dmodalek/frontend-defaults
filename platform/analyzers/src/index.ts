// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
