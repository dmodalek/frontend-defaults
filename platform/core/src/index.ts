// utility and core types
export { Constructable } from './types/Constructable';
export { IPackage } from './types/Package';
export { ValidationSeverityLevel, ValidationResult } from './utils/validation';

// filesystem utilities
export {
    getFileContents,
    getJSON,
    mergeJSON,
    fileExists,
    directoryExists,
    findFilesByPattern,
    displayPath,
} from './utils/fs';

// version utilities
export {
    DependencyInstallation,
    getDependencyInstallation,
    getInstalledDependencyVersion,
    checkDependencyUpdate,
    getLatestLTSVersion,
    checkNewerLTSVersion,
} from './utils/version';

// package utilities
export {
    getPackageJSON
} from './utils/package';

// templating utilities
export { template } from './utils/template';

// runtime functions
export { analyze } from './utils/analyze';

// TODO: clean those imports --> not needed anymore
export * from './Context';
export * from './ProjectAnalyzer';
export * from './ProjectValidator';
export * from './ProjectPatcher';

export * from './Analyzer';
export * from './Patch';
export * from './Validation';
