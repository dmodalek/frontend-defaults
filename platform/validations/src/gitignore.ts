import { GitIgnoreAnalyzerResult } from '@namics/frontend-defaults-platform-analyzers';
import { ValidationResult, ValidationSeverityLevel } from '@namics/frontend-defaults-platform-core';

export const gitIgnoreValidation = async (
	cwd: string,
	analytics: GitIgnoreAnalyzerResult
): Promise<ValidationResult[]> => {
	if (analytics.gitignore) {
		// TODO: gitignore is available, so we need to validate the contents
		return [];
	}

	return [
		{
			level: ValidationSeverityLevel.error,
			message: 'Git ignore missing',
			source: 'gitIgnoreValidation',
		},
	];
};
