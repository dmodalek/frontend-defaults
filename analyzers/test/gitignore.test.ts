import { ProjectAnalyzer } from '@namics/frontend-defaults-core';
import { getFixturePath } from './suite';
import { GitignoreAnalyzerResult, GitignoreAnalzyer } from '../src/gitignore';

const FIXTURE_INSTALLED = getFixturePath('gitignore-project');
const FIXTURE_NOT_INSTALLED = getFixturePath('default-project'); // has no .gitignore

const getFixtureAnalyzer = async (context: string): Promise<ProjectAnalyzer<GitignoreAnalyzerResult>> => {
	return new ProjectAnalyzer<GitignoreAnalyzerResult>({
		context,
		analyzers: [GitignoreAnalzyer],
	}).boot();
};

describe('Analyzers', () => {
	describe('GitignoreAnalyzer', () => {
		it('should not crash the ProjectAnalyzer', async () => {
			expect(async () => await getFixtureAnalyzer(FIXTURE_INSTALLED)).not.toThrow();
		});

		it('should analyze a project with gitignore correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE_INSTALLED);

			expect(analyzer.analytics.gitignore).toEqual(true);
			expect(analyzer.analytics).toMatchSnapshot();
		});

		it('should analyze a project without gitignore correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE_NOT_INSTALLED);

			expect(analyzer.analytics.gitignore).toEqual(false);
			expect(analyzer.analytics).toMatchSnapshot();
		});
	});
});
