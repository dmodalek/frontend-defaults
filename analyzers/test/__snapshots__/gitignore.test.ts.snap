// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers GitignoreAnalyzer should analyze a project with gitignore correctly 1`] = `
Object {
  "gitignore": true,
}
`;

exports[`Analyzers GitignoreAnalyzer should analyze a project without gitignore correctly 1`] = `
Object {
  "gitignore": false,
}
`;
